File: 211 Add and Search Word/add_and_search_word.cs
class DictNode	// Author: Xinyu Chen
{
    public const int NUM_LETTERS = 26;
    public bool isWord;
    public DictNode[] children;
    
    public DictNode()
    {
        isWord = false;
        children = new DictNode[NUM_LETTERS];
    }
}
public class WordDictionary
{
    private DictNode root;
    public WordDictionary()
    {
        root = new DictNode();
    }
    // Adds a word into the data structure.
    public void AddWord(String word)
    {
        DictNode node = root;
        foreach (var c in word)
        {
            int index = c - 'a';
            if (node.children[index] == null)
            {
                node.children[index] = new DictNode();
            }
            node = node.children[index];
        }
        node.isWord = true;
    }

    // Returns if the word is in the data structure. A word could
    // contain the dot character '.' to represent any one letter.
    public bool Search(string word)
    {
        return search(word, 0, root);
    }
    private bool search(string word, int pos, DictNode node)
    {
        if (pos == word.Length) return node.isWord;
        if (word[pos] == '.')
        {
            foreach (var n in node.children)
            {
                if (n == null) continue;
                if (search(word, pos + 1, n)) return true;
            }
            return false;
        }
        int index = word[pos] - 'a';
        if (node.children[index] == null) return false;
        return search(word, pos + 1, node.children[index]);
    }
}

// Your WordDictionary object will be instantiated and called as such:
// WordDictionary wordDictionary = new WordDictionary();
// wordDictionary.AddWord("word");
// wordDictionary.Search("pattern");

File: 211 Add and Search Word/add_and_search_word.cpp
class DictNode {
public:
    bool is_word;
    DictNode *children[26];
    DictNode() {
        is_word = false;
        for (int i = 0; i < 26; i++) children[i] = nullptr;
    }
};

class WordDictionary {
public:

    // Adds a word into the data structure.
    void addWord(string word) {
        DictNode *curr = &root;
        for (auto c : word) {
            int index = c - 'a';
            if (curr->children[index] == nullptr) {
                curr->children[index] = new DictNode();
            }
            curr = curr->children[index];
        }
        curr->is_word = true;
    }

    // Returns if the word is in the data structure. A word could
    // contain the dot character '.' to represent any one letter.
    bool search(string word) {
        return dfs(word, 0, &root);
    }
    
    bool dfs(const string& word, int pos, DictNode *node) {
        if (pos == word.size()) return node->is_word;
        bool res = false;
        if (word[pos] == '.') {
            for (int i = 0; i < 26; i++) {
                if (node->children[i] == nullptr) continue;
                res |= dfs(word, pos + 1, node->children[i]);
                if (res) return true;
            }
            return false;
        }
        int index = word[pos] - 'a';
        if (node->children[index] == nullptr) return false;
        return dfs(word, pos + 1, node->children[index]);
    }
private:
    DictNode root;
}; // Author: XC

// Your WordDictionary object will be instantiated and called as such:
// WordDictionary wordDictionary;
// wordDictionary.addWord("word");
// wordDictionary.search("pattern");

// Another method

class WordDictionary {
public:

    // Adds a word into the data structure.
    void addWord(string word) {
        hash[word.size()].insert(word);
    }

    // Returns if the word is in the data structure. A word could
    // contain the dot character '.' to represent any one letter.
    bool search(string word) {
        int len = word.size();
        if (hash.find(len) == hash.end()) return false;
        auto words = hash[word.size()];
        for (auto w : words) {
            bool flag = true;
            for (int i = 0; i < word.size(); i++) {
                if (word[i] != '.' && word[i] != w[i]) {
                    flag = false;
                    break;
                }
            }
            if (flag) return true;
        }
        return false;
    }

private:
    unordered_map<int, set<string>> hash;
}; // Author: XC

// Your WordDictionary object will be instantiated and called as such:
// WordDictionary wordDictionary;
// wordDictionary.addWord("word");
// wordDictionary.search("pattern");
